blockchain/vm/contracts.go

```diff
@@ -825,3 +862,501 @@ func (c *validateSender) validateSender(input []byte, picker types.AccountKeyPic
+	// Encode the G2 point into 256 bytes
+	return encodePointG2(r), nil
+}
+
+// bls12381G2MultiExp implements EIP-2537 G2MultiExp precompile.
+type bls12381G2MultiExp struct{}
+
+// GetRequiredGasAndComputationCost returns the gas required to execute the pre-compiled contract.
+func (c *bls12381G2MultiExp) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
+	// Calculate G2 point, scalar value pair length
+	k := len(input) / 288
+	if k == 0 {
+		// Return 0 gas for small input length
+		return 0, 0
+	}
+	// Lookup discount value for G2 point, scalar value pair length
+	var discount uint64
+	if dLen := len(params.Bls12381MultiExpDiscountTable); k < dLen {
+		discount = params.Bls12381MultiExpDiscountTable[k-1]
+	} else {
+		discount = params.Bls12381MultiExpDiscountTable[dLen-1]
+	}
+	// Calculate gas and return the result
+	return (uint64(k) * params.Bls12381G2MulGas * discount) / 1000,
+		710 * (uint64(k) * params.Bls12381G2MulGas * discount) / 1000 / 16
+	// 710 = opAddCC(150) + opMulCC(200) + opAddmodCC(360)
+	// 16 = opAddGas(3) + opMulGas(5) + opMulAddmod(8)
+}
+
+func (c *bls12381G2MultiExp) Run(input []byte, contract *Contract, evm *EVM) ([]byte, error) {
+	// Implements EIP-2537 G2MultiExp precompile logic
+	// > G2 multiplication call expects `288*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes).
+	// > Output is an encoding of multiexponentiation operation result - single G2 point (`256` bytes).
+	k := len(input) / 288
+	if len(input) == 0 || len(input)%288 != 0 {
+		return nil, errBLS12381InvalidInputLength
+	}
+	points := make([]bls12381.G2Affine, k)
+	scalars := make([]fr.Element, k)
+
+	// Decode point scalar pairs
+	for i := 0; i < k; i++ {
+		off := 288 * i
+		t0, t1, t2 := off, off+256, off+288
+		// Decode G2 point
+		p, err := decodePointG2(input[t0:t1])
+		if err != nil {
+			return nil, err
+		}
+		// 'point is on curve' check already done,
+		// Here we need to apply subgroup checks.
+		if !p.IsInSubGroup() {
+			return nil, errBLS12381G2PointSubgroup
+		}
+		points[i] = *p
+		// Decode scalar value
+		scalars[i] = *new(fr.Element).SetBytes(input[t1:t2])
+	}
+
+	// Compute r = e_0 * p_0 + e_1 * p_1 + ... + e_(k-1) * p_(k-1)
+	r := new(bls12381.G2Affine)
+	r.MultiExp(points, scalars, ecc.MultiExpConfig{})
+
+	// Encode the G2 point to 256 bytes.
+	return encodePointG2(r), nil
+}
+
+// bls12381Pairing implements EIP-2537 Pairing precompile.
+type bls12381Pairing struct{}
+
+// GetRequiredGasAndComputationCost returns the gas required to execute the pre-compiled contract.
+func (c *bls12381Pairing) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
+	k := uint64(len(input) / 384)
+	return params.Bls12381PairingBaseGas + k*params.Bls12381PairingPerPairGas,
+		params.Bls12381PairingBaseComputationCost + (k-1)*params.Bls12381PairingPerPairComputationCost
```

Comment:
Why k-1 here?

can we make it consistent to the gas formula of `params.Bls12381PairingBaseGas + k*params.Bls12381PairingPerPairGas`?

```
Bls12381PairingBaseComputationCost = Bls12381PairingBaseGas * 710/16(?)
Bls12381PairingPerPairComputationCost = Bls12381PairingPerPairGas * 710/16(?)
so that
computationcost = Bls12381PairingBaseComputationCost + k*Bls12381PairingPerPairComputationCost
```

Reply:
Fixed https://github.com/kaiachain/kaia/pull/111/commits/66e1d18b02dd1004680a42fe1cb4a5360fa0f567